### PR TITLE
Run CI in mgbuilder containers

### PR DESCRIPTION
**Pull request by MarkoBarisic**
[Parent issue](https://github.com/memgraph/infra/issues/137)

### TODOs
- [x] deploy mgbuilders with toolchain v5
  - [x] add new OSs
  - [x] keep deprecated mgbuilders with toolchain v4
  - [x] bring mgbuilder up and down on each build
  - [x] add rust and cargo
- [x] make diff workflow run entirely inside an mgbuilder container
  - [x] jobs
    - [x] community
    - [x] release
      - check e2e
    - [x] debug
    - [x] code analysis
    - [x] benchmark
    - [x] jepsen
  - [x] tests
    - [x] unit
    - [x] gql-behave
    - [x] stress-plain
    - [x] stress-ssl
    - [x] durability
    - [x] drivers
    - [x] integration 
    - [x] leftover-CTest
    - [x] ccpcheck and clang format
    - [x] clang-tidy
    - [x] code-coverage
    - [x] unit-coverage
    - [x] code-analysis
    - [x] macro-benchmark
    - [x] mgbench
    - [x] upload-to-bench-graph
    - [x] e2e
      - stream tests have been disabled, this needs to be reverted in a future PR
    - [x] jepsen
- [x] Fix gql behave issue -> patch -> [gql_behave_fix.patch](https://github.com/memgraph/memgraph/files/14392783/gql_behave_fix.patch)
- [x] Add `--ci` under `init` that ignores installing `pre_commit` python package -> we'll improve that setup later

### CODE REVIEW

File: .github/workflows/testflow.yml [hunk 1]
```diff
@@ -0,0 +1,631 @@
...
+      - name: Set up repository
+        uses: actions/checkout@v4
+        with:
+          # Number of commits to fetch. `0` indicates all history for all
+          # branches and tags. (default: 1)
+          fetch-depth: 0
+
+      - name: Spin up mgbuild container
+        run: |
+          ./release/package/run2.sh \
```
Comment (hal-eisen-MG) [inline .github/workflows/testflow.yml:38]: run2.sh needs a better (more meaningful) name

File: .github/workflows/testflow.yml [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+          --toolchain $TOOLCHAIN \
+          --os $OS \
+          --arch $ARCH \
+          --threads $THREADS \
+          --enterprise-license $MEMGRAPH_ENTERPRISE_LICENSE \
+          --organization-name $MEMGRAPH_ORGANIZATION_NAME \
+          test-memgraph unit
+
+      # - name: Ensure Kafka and Pulsar are up
```
Comment (hal-eisen-MG) [inline .github/workflows/testflow.yml:357]: Comment back in, and change if `if FALSE` mechanism, with comment explaining we will come back to fix these

File: .github/workflows/testflow.yml [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+        run: |
+          ./release/package/run2.sh \
+          --toolchain $TOOLCHAIN \
+          --os $OS \
+          --arch $ARCH \
+          --enterprise-license $MEMGRAPH_ENTERPRISE_LICENSE \
+          --organization-name $MEMGRAPH_ORGANIZATION_NAME \
+          test-memgraph e2e
+
+      # - name: Ensure Kafka and Pulsar are down
```
Comment (hal-eisen-MG) [inline .github/workflows/testflow.yml:374]: Ditto

File: release/package/run2.sh [hunk 4]
```diff
@@ -0,0 +1,587 @@
...
+
+)
+DEFAULT_THREADS=0
+DEFAULT_ENTERPRISE_LICENSE=""
+DEFAULT_ORGANIZATION_NAME="memgraph"
+# TODO(deda): 
+# * create toolchain + OS combinations
+# * add option for selecting server for push and pull
+print_help () {
+  echo -e "\nUsage: ./run.sh [OPTIONS] COMMAND [COMMAND OPTIONS]"
```
Comment (hal-eisen-MG) [inline release/package/run2.sh:48]: Script name mismatch

Comment (hal-eisen-MG) [reply release/package/run2.sh:48]: Maybe leverage the `$0` magic bash variable?

File: release/package/run2.sh [hunk 5]
```diff
@@ -0,0 +1,587 @@
...
+  echo -e "  --remove                      Remove the stopped mgbuild container"
+
+  echo -e "\ncopy options:"
+  echo -e "  --binary                      Remove the stopped mgbuild container"
+  echo -e "  --package                     Remove the stopped mgbuild container"
+
+  echo -e "\npush options:"
+  echo -e "  -p, --password string         Specify password for docker login"
+  echo -e "  -u, --username string         Specify username for docker login"
+}
```
Comment (hal-eisen-MG) [inline release/package/run2.sh:93]: Print a few example usages

File: release/package/run2.sh [hunk 6]
```diff
@@ -0,0 +1,587 @@
...
+  # Required here because Docker build container can't access remote.
+  cd "$PROJECT_ROOT"
+  if [[ "$(git rev-parse --abbrev-ref HEAD)" != "master" ]]; then
+      git fetch origin master:master
+  fi
+
+  # Ensure we have a clean build directory
+  docker exec -u mg "$build_container" bash -c "rm -rf $MGBUILD_ROOT_DIR && mkdir -p $MGBUILD_ROOT_DIR"
+
+  # TODO(gitbuda): Revisit copying the whole repo -> makese sense under CI.
```
Comment (hal-eisen-MG) [inline release/package/run2.sh:236]: Is this still relevant?

File: release/package/run2.sh [hunk 7]
```diff
@@ -0,0 +1,587 @@
...
+  docker exec -u mg "$build_container" bash -c "cd $MGBUILD_ROOT_DIR && git remote set-url origin https://github.com/memgraph/memgraph.git"
+
+  # Define cmake command
+  local cmake_cmd="cmake $build_type_flag $arm_flag $community_flag $telemetry_id_override_flag $coverage_flag $asan_flag $ubsan_flag .."
+  docker exec -u mg "$build_container" bash -c "cd $container_build_dir && $ACTIVATE_TOOLCHAIN && $cmake_cmd"
+  
+  # ' is used instead of " because we need to run make within the allowed
+  # container resources.
+  # shellcheck disable=SC2016
+  if [[ "$threads" == 0 ]]; then
```
Comment (hal-eisen-MG) [inline release/package/run2.sh:264]: Add a comment about how this is a workaround for MacOS

File: .github/workflows/testflow.yml (hunk 1 above)
Comment (MarkoBarisic) [reply .github/workflows/testflow.yml]: Renamed to mgbuild.sh
Reasoning: the script is intended to interact with mgbuild containers, with all the features

File: src/storage/v2/replication/replication_client.cpp
[the remark this answers is not in the document]
Comment (MarkoBarisic) [reply src/storage/v2/replication/replication_client.cpp:1]: @andrejtonev created a PR from which I cherry picked this commit, it was a quick fix.

File: .github/workflows/diff.yaml [hunk 8]
```diff
@@ -168,58 +213,95 @@ jobs:
...
+      - name: Spin up mgbuild container
         run: |
-          # Activate toolchain.
-          source /opt/toolchain-v4/activate
-
-          # Initialize dependencies.
-          ./init
+          ./release/package/mgbuild.sh \
+          --toolchain $TOOLCHAIN \
+          --os $OS \
```
Comment (hal-eisen-MG) [inline .github/workflows/diff.yaml:220]: Fix whitespace @MarkoBarisic

**Review by hal-eisen-MG — APPROVED** [review]
LGTM

**Pull request merged**